tests/testthat/test-utility-extra.R

```diff
@@ -0,0 +1,400 @@
+  names(dimnames(res$cw)) = c('', '')
+  testthat::expect_equal(res, ref_list)
+})
+
+testthat::test_that(
+  "Intersecting cones works (when all pedestrians are seen in front)", {
+  ref = m4ma::iCones_r(p1, a_front, p2[-n, , drop = FALSE], r, objects)
+  res = m4ma::iCones_rcpp(p1, a_front, p2[-n, , drop = FALSE], r, objects)
+  testthat::expect_equal(res, ref)
+})
+
+testthat::test_that(
+  "Intersecting cones works (when single pedestrian)", {
+    ref = m4ma::iCones_r(p1, a_front, matrix(, 0, 2), r, objects)
+    res = m4ma::iCones_rcpp(p1, a_front, matrix(, 0, 2), r, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one other pedestrian)", {
+    ref = m4ma::iCones_r(p1, a_front, p2[2, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, a_front, p2[2, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when no intersection)", {
+    ref = m4ma::iCones_r(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting and seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting but not seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front,
+                         p2[3, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_not_front,
+                            p2[3, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when all pedestrians are occluded in front)", {
+    ref = m4ma::iCones_r(p1, a_front,
+                         p2[-n, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_front,
+                            p2[-n, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+  })
+
+testthat::test_that("Intersecting cones to cells transformation works", {
+  # Make test independent of iCones
+  cones = c(rnorm(6))
+  names(cones) = as.character(1:6)
+  v = 0.5
+  ref = m4ma::iCones2Cells_r(cones, v)
+  res = m4ma::iCones2Cells_rcpp(cones, v)
+  testthat::expect_equal(res, ref)
+})
+
+testthat::test_that(
+  "Blocked angle works (when intersecting cones and seen)", {
+    ref = m4ma::blockedAngle_r(n, state, p_pred, objects)
+    res = m4ma::blockedAngle_rcpp(n-1, state, p_pred, objects)
+    testthat::expect_equal(res, if (length(ref) == 0) unname(ref) else ref)

```

Comment:
When there is only one case, please consider in R the opportunity to use ifelse() rather than if() else() as the former is vectorised, and it is more concise.

So:

ifelse(length(ref) == 0, unname(ref), ref)

Reply:
I tried this and it does not work because it is indeed vectorized so it checks every element of ref instead of the whole vector for `length()`. It's not pretty, but I will leave it as it is.

Reply:
fair enough!